alpha-commons/src/main/java/at/ac/tuwien/kr/alpha/commons/programs/ReificationHelper.java

```diff
@@ -0,0 +1,455 @@
+	static final ConstantTerm<String> ATOM_TYPE_COMPARISON = Terms.newSymbolicConstant("comparison");
+	static final ConstantTerm<String> ATOM_TYPE_EXTERNAL = Terms.newSymbolicConstant("external");
+	static final ConstantTerm<String> ATOM_TYPE_AGGREGATE = Terms.newSymbolicConstant("aggregate");
+
+	// Constants describing comparison operators.
+	static final ConstantTerm<String> CMP_OP_EQ = Terms.newSymbolicConstant("eq");
+	static final ConstantTerm<String> CMP_OP_NE = Terms.newSymbolicConstant("ne");
+	static final ConstantTerm<String> CMP_OP_LE = Terms.newSymbolicConstant("le");
+	static final ConstantTerm<String> CMP_OP_LT = Terms.newSymbolicConstant("lt");
+	static final ConstantTerm<String> CMP_OP_GE = Terms.newSymbolicConstant("ge");
+	static final ConstantTerm<String> CMP_OP_GT = Terms.newSymbolicConstant("gt");
+
+	// Constants describing aggregate functions.
+	static final ConstantTerm<String> AGGREGATE_FUNCTION_COUNT = Terms.newSymbolicConstant("count");
+	static final ConstantTerm<String> AGGREGATE_FUNCTION_SUM = Terms.newSymbolicConstant("sum");
+	static final ConstantTerm<String> AGGREGATE_FUNCTION_MIN = Terms.newSymbolicConstant("min");
+	static final ConstantTerm<String> AGGREGATE_FUNCTION_MAX = Terms.newSymbolicConstant("max");
+
+	// Constants describing term types.
+	static final ConstantTerm<String> TERM_TYPE_CONSTANT = Terms.newSymbolicConstant("constant");
+	static final ConstantTerm<String> TERM_TYPE_VARIABLE = Terms.newSymbolicConstant("variable");
+	static final ConstantTerm<String> TERM_TYPE_ARITHMETIC = Terms.newSymbolicConstant("arithmetic");
+	static final ConstantTerm<String> TERM_TYPE_FUNCTION = Terms.newSymbolicConstant("function");
+
+	private static final Map<ComparisonOperator, ConstantTerm<String>> CMP_OPS = new HashMap<>();
+	private static final Map<AggregateFunctionSymbol, ConstantTerm<String>> AGG_FUNCS = new HashMap<>();
+	private static final Map<Class<?>, String> TERM_TYPES = new HashMap<>();
+
+	static {
+		CMP_OPS.put(ComparisonOperators.EQ, CMP_OP_EQ);
+		CMP_OPS.put(ComparisonOperators.NE, CMP_OP_NE);
+		CMP_OPS.put(ComparisonOperators.LE, CMP_OP_LE);
+		CMP_OPS.put(ComparisonOperators.LT, CMP_OP_LT);
+		CMP_OPS.put(ComparisonOperators.GE, CMP_OP_GE);
+		CMP_OPS.put(ComparisonOperators.GT, CMP_OP_GT);
+
+		AGG_FUNCS.put(AggregateFunctionSymbol.COUNT, AGGREGATE_FUNCTION_COUNT);
+		AGG_FUNCS.put(AggregateFunctionSymbol.SUM, AGGREGATE_FUNCTION_SUM);
+		AGG_FUNCS.put(AggregateFunctionSymbol.MIN, AGGREGATE_FUNCTION_MIN);
+		AGG_FUNCS.put(AggregateFunctionSymbol.MAX, AGGREGATE_FUNCTION_MAX);
+
+		TERM_TYPES.put(String.class, "string");
+		TERM_TYPES.put(Integer.class, "integer");
+	}
+
+	private final Supplier<ConstantTerm<?>> idProvider;
+
+	private final Map<Predicate, ConstantTerm<?>> reifiedPredicates = new HashMap<>();
+
+	public ReificationHelper(Supplier<ConstantTerm<?>> idProvider) {
+		this.idProvider = idProvider;
```

Comment:
I created a new interface `IdGenerator<T>` which has an explicit contract stating that fresh IDs are generated with each call. This is now used instead of `Supplier`.